doc/man3/EVP_MAC.pod

```diff
@@ -0,0 +1,235 @@
+=pod
+
+=head1 NAME
+
+EVP_MAC, EVP_MAC_CTX, EVP_MAC_CTX_new, EVP_MAC_CTX_new_id, EVP_MAC_CTX_free,
+EVP_MAC_CTX_copy, EVP_MAC_CTX_mac, EVP_MAC_reset, EVP_MAC_size, EVP_MAC_init,
+EVP_MAC_update, EVP_MAC_final, EVP_MAC_ctrl, EVP_MAC_ctrl_str, EVP_MAC_str2ctrl,
+EVP_MAC_hex2ctrl, EVP_MAC_nid, EVP_MAC_name, EVP_add_mac, EVP_get_macbyname,
+EVP_get_macbynid, EVP_get_macbyobj - EVP MAC routines
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ typedef struct evp_mac_st EVP_MAC;
+ typedef struct evp_mac_ctx_st EVP_MAC_CTX;
+
+ EVP_MAC_CTX *EVP_MAC_CTX_new(const EVP_MAC *mac);
+ EVP_MAC_CTX *EVP_MAC_CTX_new_id(int nid);
+ void EVP_MAC_CTX_free(EVP_MAC_CTX *ctx);
+ int EVP_MAC_CTX_copy(EVP_MAC_CTX *dest, EVP_MAC_CTX *src);
+ const EVP_MAC *EVP_MAC_CTX_mac(EVP_MAC_CTX *ctx);
+ int EVP_MAC_reset(EVP_MAC_CTX *ctx);
+ size_t EVP_MAC_size(EVP_MAC_CTX *ctx);
+ int EVP_MAC_init(EVP_MAC_CTX *ctx);
+ int EVP_MAC_update(EVP_MAC_CTX *ctx, const unsigned char *data, size_t datalen);
+ int EVP_MAC_final(EVP_MAC_CTX *ctx, unsigned char *out, size_t *poutlen);
+ int EVP_MAC_ctrl(EVP_MAC_CTX *ctx, int cmd, ...);
+ int EVP_MAC_ctrl_str(EVP_MAC_CTX *ctx, const char *type, const char *value);
+ int EVP_MAC_str2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_hex2ctrl(EVP_MAC_CTX *ctx, int cmd, const char *value);
+ int EVP_MAC_nid(const EVP_MAC *mac);
+ const char *EVP_MAC_name(const EVP_MAC *mac);
+ int EVP_add_mac(const EVP_MAC *mac);
+ const EVP_MAC *EVP_get_macbyname(const char *name);
+ const EVP_MAC *EVP_get_macbynid(int nid);
+ const EVP_MAC *EVP_get_macbyobj(const ASN1_OBJECT *o);
+
+=head1 DESCRIPTION
+
+These types and functions help the application to calculate MACs of
+different types and with different underlying algorithms if there are
+any.
+
+=head2 Types
+
+B<EVP_MAC> is a type that holds the implementation of a MAC.
+
+B<EVP_MAC_CTX> is a context type that holds internal information used
+to compute MACs from the input.
+
+=head2 Context manipulation functions
+
+EVP_MAC_CTX_new() creates a new context for the MAC type C<mac>.
+EVP_MAC_CTX_new_id() creates a new context for the numerical MAC
+identity <nid>.  The created context can then be used with most other
+functions described here.
+
+EVP_MAC_CTX_free() frees the contents of the context as well as the
+context itself.  B<NULL> is a valid parameter, for which this function
+is a no-op.
+
+EVP_MAC_CTX_copy() makes a deep copy of the C<src> context to the
+C<dest> context.  The C<dest> context I<must> have been created before
+calling this function.
```

Comment:
Maybe a dup() function is more useful?

Reply:
```
: ; grep dup include/openssl/evp.h 
EVP_MD *EVP_MD_meth_dup(const EVP_MD *md);
EVP_CIPHER *EVP_CIPHER_meth_dup(const EVP_CIPHER *cipher);
EVP_PKEY_CTX *EVP_PKEY_CTX_dup(EVP_PKEY_CTX *ctx);
: levitte@lapdog:~/gitwrk/openssl.net/official/master
 * this file except in compliance with the License.  You can obtain a copy
int EVP_MD_meth_set_copy(EVP_MD *md, int (*copy)(EVP_MD_CTX *to,
int (*EVP_MD_meth_get_copy(const EVP_MD *md))(EVP_MD_CTX *to,
/* cipher has its own additional copying logic */
int EVP_CIPHER_CTX_copy(EVP_CIPHER_CTX *out, const EVP_CIPHER_CTX *in);
__owur int EVP_MD_CTX_copy_ex(EVP_MD_CTX *out, const EVP_MD_CTX *in);
__owur int EVP_MD_CTX_copy(EVP_MD_CTX *out, const EVP_MD_CTX *in);
int EVP_ENCODE_CTX_copy(EVP_ENCODE_CTX *dctx, EVP_ENCODE_CTX *sctx);
int EVP_PKEY_copy_parameters(EVP_PKEY *to, const EVP_PKEY *from);
void EVP_PKEY_asn1_copy(EVP_PKEY_ASN1_METHOD *dst,
                             int (*param_copy) (EVP_PKEY *to,
void EVP_PKEY_meth_copy(EVP_PKEY_METHOD *dst, const EVP_PKEY_METHOD *src);
void EVP_PKEY_meth_set_copy(EVP_PKEY_METHOD *pmeth,
                            int (*copy) (EVP_PKEY_CTX *dst,
void EVP_PKEY_meth_get_copy(const EVP_PKEY_METHOD *pmeth,
                            int (**pcopy) (EVP_PKEY_CTX *dst,
```

(an interesting thing to note is that EVP_CIPHER has a copy function for its context, but calls the method function pointer `dup` :wink:)

I think I'll stick with the copy pattern...

Reply:
I really dislike the copy pattern. I'm sure most of those copy functions have various bugs in them and are very hard to get correct, and probably none of them are correct. This is part of the reusing pattern that I want to get rid of.

Reply:
The options for me are either providing a dup function, or removing this function.

Reply:
I think I recall seeing calls to copy when running through tests. I'll do a run with a profiler to see if that's still the case or if it went away somehow.

Reply:
As it is right now, we cannot get away from a copy function, because there are variants of key passing down to an MAC (a CMAC in my testing case) from `EVP_DigestSignInit` that requires copying a `EVP_MAC_CTX`.  `EVP_PKEY_new_CMAC_key` is the function used to create that `EVP_MAC_CTX`.  So getting rid of the copy function means digging very deep in this MAC-over-PKEY stuff.  I'm not ready to do that, and *certainly* not as part of this PR.

It would of course be desirable to get rid of MAC-over-PKEY, but since it's part of the current API and something people have learned to use, we can't...  which means that for this PR, the copy function stays.